### PR TITLE
Add documentation for the compression implementation

### PR DESCRIPTION
**Pull request by pushkarnk**

### CODE REVIEW

File: Compression.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+## WebSocket Compression
+WebSocket Compression, defined by [RFC7692](https://tools.ietf.org/html/rfc7692) allows WebSocket clients to send and receive compressed data on a WebSocket connection. Compression reduces the total wire-level payload of a WebSocket connection, possibly resulting in an improved throughput.
+
+This document discusses the implementation of WebSocket Compression in [Kitura-WebSocket-NIO](https://github.com/IBM-Swift/Kitura-WebSocket-NIO), which is an implementation of the [Kitura-WebSocket](https://github.com/IBM-Swift/Kitura-WebSocket) API using [swift-nio](https://github.com/apple/swift-nio).
```
Comment (helenmasters) [inline Compression.md]: A minor question - do we usually refer to swift-nio or should we be using the name SwiftNIO here (the link to the github repo is fine either way)?

Comment (ianpartridge) [reply Compression.md]: SwiftNIO

File: Compression.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+2. [WebSocket compression - the `permessage-deflate` algorithm](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#2-the-permessage-deflate-algorithm)
+3. [An implementation of permessage-deflate based on swift-nio](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#3-an-implementation-of-permessage-deflate-based-on-swift-nio)
+4. [Developer notes](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#4-developer-notes)
+
+### 1. WebSocket extensions
+The WebSocket protocol has a provision for servers to configure protocol extensions, and for clients to request these extensions from the servers. A client notifies about extensions it is interested in through a `negotiation offer` using the `Sec-WebSocket-Extension` header. A server may or may not support the extensions requested by the client. Through a `negotiation response`, the server notifies the client of the extensions that the server agrees upon. The negotiation response offer and response may also include extension-specific parameters. Once an extension is agreed upon, the client and server must invoke the extension from their respective WebSocket implementations.
```
Comment (helenmasters) [inline Compression.md]: I found the wording "The negotiation response offer and response may also..." confusing as we were referring to `negotiation offer` and `negotiation response` previously.

File: Compression.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ - `client_no_context_takeover`, `server_no_context_takeover`
+ - `client_max_window_bits`, `server_max_window_bits`
+We shall revisit these parameters in the future sections, where we discuss their use and effects.
+
+A permessage-deflate negotiation response has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters, agreed upon by the server. The headers in the negotiation response are the final word on how the data compression/decompression will be done between the client and server. Data compressed by the client must be decompressed by the server and vice versa. The client and server must adopt the same compression/decompression configuration parameters. We shall deal with this in detail in the future sections.
+
+The specification also discusses the DEFLATE algorithm. We adopt the [zlib](https://www.zlib.net) for doing raw compression and decompression. A pair of a compressor and a decompressor must be set up at both the ends of the connection. The server's decompressor decompresses messages compressed by the client's compressor and vice versa.
```
Comment (helenmasters) [inline Compression.md]: It reads better as "the [zlib](https://www.zlib.net) compression library" but are we trying to say "In our implementation we have utilized?(or made use of?) the zlib compression library for doing raw compression and decompression."

File: Compression.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+WebSocket compression is a WebSocket extension.
+
+### 2. WebSocket Compression: the permessage-deflate algorithm
+Permessage-deflate is a WebSocket extension defined by [RFC7692](https://tools.ietf.org/html/rfc7692) which provides a specification for the compression functionality. It defines the negotiation process and a compression algorithm called `DEFLATE`. 
+
+A `permessage-deflate` negotiation happens during the upgrade request from HTTP to WebSocket. A `permessage-deflate` negotiation offer has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters. There are four extension parameters defined for WebSocket compression:
+ - `client_no_context_takeover`, `server_no_context_takeover`
+ - `client_max_window_bits`, `server_max_window_bits`
+We shall revisit these parameters in the future sections, where we discuss their use and effects.
```
Comment (helenmasters) [inline Compression.md]: You need an extra new line in front of this sentence as it currently appears at the end of the list.

File: Compression.md (hunk 3 above)
Comment (helenmasters) [reply Compression.md]: "A pair of a compressor and a decompressor must be set up at both the end of the connection." can we reword this to something like "A pair, consisting of a compressor and a decompressor, must be set up at both ends of the connection."

File: Compression.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+The specification also discusses the DEFLATE algorithm. We adopt the [zlib](https://www.zlib.net) for doing raw compression and decompression. A pair of a compressor and a decompressor must be set up at both the ends of the connection. The server's decompressor decompresses messages compressed by the client's compressor and vice versa.
+
+### 3. An implementation of permessage-deflate based on swift-nio
+The [swift-nio](https://github.com/apple/swift-nio) framework provides an API which enables HTTP/WebSocket server implementations view the processing of data, read from or written to sockets, as a sequence of transformations that happen through a pipeline of handlers. An active connection is represented by a `Channel`. Data which is read from, or written to, a channel moves through a `ChannelPipeline` of inbound and outbound `ChannelHandlers`. An `EventLoop` is associated with every `Channel`. An `EventLoop`s is a thread-safe abstraction of a thread and provides features for asynchronous code execution using `EventLoopFutures` and `EventLoopPromises`.
```
Comment (helenmasters) [inline Compression.md]: Can we reword the first bit of this sentence...?
"The [SwiftNIO](https://github.com/apple/swift-nio) framework provides an API which enables HTTP/WebSocket server implementations to view the processing of data, that has been read from or written to sockets, as...."

Comment (helenmasters) [inline Compression.md]: An `EventLoop`s - you need to lose the 's'!

File: Compression.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+WebSocket compression is a WebSocket extension.
+
+### 2. WebSocket Compression: the permessage-deflate algorithm
+Permessage-deflate is a WebSocket extension defined by [RFC7692](https://tools.ietf.org/html/rfc7692) which provides a specification for the compression functionality. It defines the negotiation process and a compression algorithm called `DEFLATE`. 
+
+A `permessage-deflate` negotiation happens during the upgrade request from HTTP to WebSocket. A `permessage-deflate` negotiation offer has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters. There are four extension parameters defined for WebSocket compression:
```
Comment (helenmasters) [inline Compression.md]: Given there is quite a lot of text in here it might make sense to have a "`permessage-deflate` negotiation offer" minor heading and "`permessage-deflate` negotiation response" minor heading to make this section a little easier to read. It's just a suggestion though.

File: Compression.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    *(Kitura/WebSocket app)*  
+    \|  
+    `HTTPEncoder`  
+    \|  
+    `NIOSSLServerHandler`  
+    \|  
+    *(Operating System)*   
+
+[HTTPDecoder](https://apple.github.io/swift-nio/docs/current/NIOHTTP1/Classes/HTTPDecoder.html) and [HTTPResponseEncoder](https://apple.github.io/swift-nio/docs/current/NIOHTTP1/Classes/HTTPResponseEncoder.html) convert bytes to HTTP requests, and responses to bytes, respectively. [NIOSSLServerHandler](https://apple.github.io/swift-nio-ssl/docs/current/NIOSSL/Classes/NIOSSLServerHandler.html) is a duplex handler (both inbound and outbound) used to decrypt and encrypt data on a secure connection. The [HTTPRequestHandler](https://github.com/IBM-Swift/Kitura-NIO/blob/master/Sources/KituraNet/HTTP/HTTPRequestHandler.swift) is used to invoke Kitura's router.
```
Comment (helenmasters) [inline Compression.md]: Why is the HTTPRequestHandler link the only one that doesn't link to API doc?

File: Compression.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+An upgrade to WebSocket causes swift-nio to alter the above pipeline in these ways:
+ - the `HTTPDecoder` and `HTTPRespobseEncoder` (and other HTTP related handlers) are removed from the pipeline
```
Comment (helenmasters) [inline Compression.md]: `HTTPRespobseEncoder`?

File: Compression.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    \|  
+   `WebSocketFrameEncoder`      
+    \|  
+   `NIOSSLServerHandler`  
+    \|  
+   *(Operating System)*    
+ 
+ [PermessageDeflateCompressor](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/blob/master/Sources/KituraWebSocket/PermessageDeflateCompressor.swift) is an outbound handler used to compress outbound WebSocket messages. [PermessageDeflateDecompressor](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/blob/master/Sources/KituraWebSocket/PermessageDeflateDecompressor.swift) is an inbound handler used to decompress inbound WebSocket messages. Every WebSocket connection where a valid `permessage-deflate` compression was negotiated, gets its own (`PermessageDeflateCompressor`, `PermessageDeflateDecompressor`) pair. 
+ 
+ With this setup, all the inbound data first passes through NIO's `WebSocketDecoder` where the WebSocket frames are built. It then moves into the `PermessageDeflateDecompressor` where multiple frames comprising a message are accumulated and decompressed using `zlib`'s inflater. Subsequently, the decompressed messages are moved to the `WebSocketConnection` handler.
```
Comment (helenmasters) [inline Compression.md]: It would be nice to have an API link for `WebSocketDecoder`.
I've also lost track as to what we're referring to here by NIO's, do you mean SwiftNIO's or something else?

File: Compression.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+#### 3.1 Compressor implementation
+The compressor is called `PermessageDeflateCompressor`. It is a [ChannelOutboundHandler](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/ChannelOutboundHandler.html).
+
+`ChannelOutboundHandler`'s `write(context:data:promise)` method implemented here gets invoked when the previous outbound handler (`WebSocketConnection`) writes data to the channel. Here, only data frames and continuation frames are processed. A WebSocket message is either available in a single data frame or a data frame followed by sequence of continuation frames. 
+
+The compressor makes sure that we have all the data pertaining to a message accumulated. Subsequently, `zlib`'s deflater is invoked and deflated data is packed into a new WebSocketFrame, which is passed to the `WebSocketFrameEncoder`.
+
+#### 3.2 Decompressor implementation
+The decompressor is, functionally, a mirror image of the compressor. It is called `PermessageDeflateDecompressor` and is [ChannelInboundHandler](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/ChannelInboundHandler.html).
```
Comment (helenmasters) [inline Compression.md]: missing 'a' in "and is ChannelInboundHandler"

File: Compression.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+`ChannelInboundHandler`'s `channelRead(context:data)` method implemented here is invoked whenever a new `WebSocketFrame`s is produced by the `WebSocketDecoder`. Similar to the compressor, the decompressor processes data and continuation frames only. All the data pertaining to a message, possibly spread across continuation frames, is accumulated and `zlib`'s inflater is invoked. The inflated data is then packed into a new `WebSocketFrame` and moved into the next handler in the pipeline.
```
Comment (helenmasters) [inline Compression.md]: Need to lose the 's' after `WebSocketFrame`

File: Compression.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+#### 3.3 Configuring the Compressor and Decompressor
+[RFC7692](https://tools.ietf.org/html/rfc7692) defines four configuration options. They are actually two pairs of option, one for the client and server each:
```
Comment (helenmasters) [inline Compression.md]: Can we change this to "They are actually two pairs of options, one each for the client and the server:"

File: Compression.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+###### 3.3.1 `client_no_context_takeover` and `server_no_context_takeover` 
+These allow the client and server to use a new zlib inflater or deflater on every message. By default we reuse the inflater and deflater instances across messages. This means the inflater and deflater are initialized only once. The memory allocation/deallocation happen only once and the history of the deflate/inflate stream can be reused. [Here](https://tools.ietf.org/html/rfc7692#section-7.2.3.2) is an example.
```
Comment (helenmasters) [inline Compression.md]: happen -> happens

File: Compression.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+In the `Kitura-WebSocket-NIO` implementation:
+ - a client can inform the server that it isn't using context takeover by sending the `client_no_context_takeover` extension parameter. The server will respond with the same `client_no_context_takeover` parameter and configure its decompressor to not use context takeover.
+ - a client can request the server to not use context takeover by sending the `server_no_context_takeover` extension parameter. The server will regard this request and configure its compressor to not use context takeover. It will add the `server_no_context_takeover` parameter in the response.
+ - by default the server will configure its compressor and decompressor to use context takeover .i.e. to reuse compression context
```
Comment (helenmasters) [inline Compression.md]: .i.e. -> i.e.

File: Compression.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+###### 3.3.2 `client_max_window_bits` and `server_max_window_bits` 
+These allow the client and server to share the LZ77 sliding window size. The default value is 15 bits which represents a window size of 32768 (2^15). It is necessary that the client's compressor and server's decompressor share the same LZ77 window size. So must it be between the server's compressor and client's decompressor.
+
+In the `Kitura-WebSocket-NIO` implementation:
+- a client can inform the server of it's compressor's LZ77 window size using the `client_max_window_bits` extension parameter. If the parameter has a valid value, the server will configure its decompressor accordingly and send the same extension parameter in the response, indicating an agreement.
```
Comment (helenmasters) [inline Compression.md]: it's -> its

File: Compression.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+In the `Kitura-WebSocket-NIO` implementation:
+ - a client can inform the server that it isn't using context takeover by sending the `client_no_context_takeover` extension parameter. The server will respond with the same `client_no_context_takeover` parameter and configure its decompressor to not use context takeover.
+ - a client can request the server to not use context takeover by sending the `server_no_context_takeover` extension parameter. The server will regard this request and configure its compressor to not use context takeover. It will add the `server_no_context_takeover` parameter in the response.
+ - by default the server will configure its compressor and decompressor to use context takeover .i.e. to reuse compression context
+
+###### 3.3.2 `client_max_window_bits` and `server_max_window_bits` 
+These allow the client and server to share the LZ77 sliding window size. The default value is 15 bits which represents a window size of 32768 (2^15). It is necessary that the client's compressor and server's decompressor share the same LZ77 window size. So must it be between the server's compressor and client's decompressor.
```
Comment (helenmasters) [inline Compression.md]: I might read better as "The client's compressor and the server's decompressor must have the same LZ77 window size. The same restriction applies to the server's compressor and the client's decompressor."

File: Compression.md [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+3. The LZ77 sliding window value must be passed as a negative parameter to deflateInit2()/inflateInit2(). This informs `zlib` that we use raw deflate streams (as against zlib streams that result with sliding window positive values). See [this](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/pull/26/commits/52614a23dbff37db18e2c0fb70e282921c0bb666).
+
+4. There's an open zlib bug with a sliding window size of 8 bits. See [this](https://github.com/madler/zlib/issues/171). There is a workaround for zlib streams. We implement a similar workaround for raw deflate streams [here](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/pull/26/commits/c9d9b44d56de7c398a526b34dacf7b4302045f73).
+
+5. Clients may negotiate for compression but send uncompressed frames. To handle this, just before decompressing, we check the RSV1 bit of the first frame to make sure it belongs to a compressed message.
+
+6. We can merge the two handlers - `PermessageDeflateCompressor` and `PermessageDeflateDecompressor` into a single duplex handler. We can also make the `WebSocketConnection` a duplex handler.
```
Comment (helenmasters) [inline Compression.md]: Do you have an example of what you mean by this?

File: Compression.md [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
```
Comment (helenmasters) [inline Compression.md]: Are there any examples we can point people at if they'd like to see a working implementation? Is there anything in Kitura-sample?

File: Compression.md [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+A `permessage-deflate` negotiation happens during the upgrade request from HTTP to WebSocket. A `permessage-deflate` negotiation offer has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters. There are four extension parameters defined for WebSocket compression:
+ - `client_no_context_takeover`, `server_no_context_takeover`
+ - `client_max_window_bits`, `server_max_window_bits`
+We shall revisit these parameters in the future sections, where we discuss their use and effects.
+
+A permessage-deflate negotiation response has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters, agreed upon by the server. The headers in the negotiation response are the final word on how the data compression/decompression will be done between the client and server. Data compressed by the client must be decompressed by the server and vice versa. The client and server must adopt the same compression/decompression configuration parameters. We shall deal with this in detail in the future sections.
```
Comment (helenmasters) [inline Compression.md]: We shall deal with this in detail in the future sections. -> "We will talk about this in detail later in this document."

File: Compression.md (hunk 1 above)
Comment (pushkarnk) [reply Compression.md]: Replaced all instances of `swift-nio` by `SwiftNIO`

File: Compression.md (hunk 2 above)
Comment (pushkarnk) [reply Compression.md]: Thanks. Changing it to `Negotiation offers and responses may also include...`

File: Compression.md (hunk 4 above)
Comment (pushkarnk) [reply Compression.md]: I agree.

File: Compression.md (hunk 7 above)
Comment (pushkarnk) [reply Compression.md]: The `HTTPRequestHandler` is not a part of the API. It's an internal class.

File: Compression.md (hunk 9 above)
Comment (pushkarnk) [reply Compression.md]: Thanks. Added API links to `WebSocketDecoder` and `WebSocketEncoder`.

File: Compression.md (hunk 16 above)
Comment (pushkarnk) [reply Compression.md]: I agree.

File: Compression.md (hunk 17 above)
Comment (pushkarnk) [reply Compression.md]: I've added a short definition of a `ChannelDuplexHandler` and also linked to its API doc (which doesn't have anything helpful, actually).

File: Compression.md (hunk 18 above)
Comment (pushkarnk) [reply Compression.md]: We don't have a great example to demonstrate compression. I've created https://github.com/IBM-Swift/Kitura-WebSocket-NIO/issues/39 to track this.

**Comment by pushkarnk** [comment]
Thanks @helenmasters for the great review! I've responded to all of your comments.
